client/core/core.go

```diff
@@ -5959,13 +5892,245 @@ func (c *Core) prepareTradeRequest(pw []byte, form *TradeForm) (*tradeRequest, e
+				return nil, newError(orderParamsErr, "cannot estimate redemption count")
+			}
+		}
+	}
+
+	if lots == 0 {
+		return nil, newError(orderParamsErr, "order quantity < 1 lot. qty = %d %s, rate = %d, lot size = %d",
+			qty, assetConfigs.baseAsset.Symbol, rate, mktConf.LotSize)
+	}
+
+	coins, redeemScripts, fundingFees, err := fromWallet.FundOrder(&asset.Order{
+		Version:       assetConfigs.fromAsset.Version,
+		Value:         fundQty,
+		MaxSwapCount:  lots,
+		MaxFeeRate:    assetConfigs.fromAsset.MaxFeeRate,
+		Immediate:     isImmediate,
+		FeeSuggestion: c.feeSuggestion(dc, assetConfigs.fromAsset.ID),
+		Options:       form.Options,
+		RedeemVersion: assetConfigs.toAsset.Version,
+		RedeemAssetID: assetConfigs.toAsset.ID,
+	})
+	if err != nil {
+		return nil, codedError(walletErr, fmt.Errorf("FundOrder error for %s, funding quantity %d (%d lots): %w",
+			assetConfigs.fromAsset.Symbol, fundQty, lots, err))
+	}
+	defer func() {
+		if _, err := c.updateWalletBalance(fromWallet); err != nil {
+			c.log.Errorf("updateWalletBalance error: %v", err)
+		}
+		if fromToken := asset.TokenInfo(assetConfigs.fromAsset.ID); fromToken != nil {
+			c.updateAssetBalance(fromToken.ParentID)
+		}
+	}()
+
+	// The coins selected for this order will need to be unlocked
+	// if the order does not get to the server successfully.
+	errCloser := dex.NewErrorCloser()
+	defer errCloser.Done(c.log)
+	errCloser.Add(func() error {
+		err := fromWallet.ReturnCoins(coins)
+		if err != nil {
+			return fmt.Errorf("Unable to return %s funding coins: %v", unbip(fromWallet.AssetID), err)
+		}
+		return nil
+	})
+
+	tradeRequest, err := c.createTradeRequest(wallets, coins, redeemScripts, dc, redeemAddr, form,
+		lots, redemptionRefundLots, fundingFees, assetConfigs, mktConf, errCloser)
+	if err != nil {
+		return nil, err
 	}
 
 	errCloser.Success()
 
 	return tradeRequest, nil
 }
 
+func (c *Core) prepareMultiTradeRequests(pw []byte, form *MultiTradeForm) ([]*tradeRequest, error) {
+	wallets, assetConfigs, dc, mktConf, err := c.prepareForTradeRequestPrep(pw, form.Base, form.Quote, form.Host, form.Sell)
+	if err != nil {
+		return nil, err
+	}
+	fromWallet, toWallet := wallets.fromWallet, wallets.toWallet
+
+	multiFunder, is := fromWallet.Wallet.(asset.MultiOrderFunder)
+	if !is {
+		return nil, newError(orderParamsErr, "fromWallet is not a MultiOrderFunder")
+	}
+
+	for _, trade := range form.Placements {
+		if trade.Rate == 0 {
+			return nil, newError(orderParamsErr, "zero rate is invalid")
+		}
+		if trade.Qty == 0 {
+			return nil, newError(orderParamsErr, "zero quantity is invalid")
+		}
+	}
+
+	redeemAddresses := make([]string, 0, len(form.Placements))
+	for range form.Placements {
+		redeemAddr, err := toWallet.RedemptionAddress()
+		if err != nil {
+			return nil, codedError(walletErr, fmt.Errorf("%s RedemptionAddress error: %w",
+				assetConfigs.toAsset.Symbol, err))
+		}
+		redeemAddresses = append(redeemAddresses, redeemAddr)
+	}
```

Comment:
If we're replacing orders en masse when the market moves, we could potentially be generating a lot of unused redemption addresses. I wonder if we need to consider the gap limit.

Reply:
I also think gap limit is a real concern, particularly with BTC.  Since the gap limit was already becoming an issue before this, we recenly made it so DCR now always [uses the wrap gap policy](https://github.com/decred/dcrdex/commit/32607367cef05aef22cb57d3fc524c7894b046ed) so we don't make big gaps.  With BTC I'm not sure what we did or can do.  I think there's a 1000 address limit with the full node wallet and 250 with native, but we should research that.

Do you think we need to make some kind of address pool for BTC to alleviate this?

Reply:
We're really close on the gap limit issues thanks to you, @buck54321.  Just a few minor comments there needing response/revision https://github.com/decred/dcrdex/pull/2368